src/System.Drawing.Common/src/System/Drawing/Icon.Unix.cs

```diff
@@ -709,77 +709,163 @@ private void InitFromStreamWithSize(Stream stream, int width, int height)
+                                if (nread != xorSize)
+                                {
+                                    string msg = string.Format("{0} data length expected {1}, read {2}", "XOR", xorSize, nread);
+                                    throw new ArgumentException(msg, nameof(stream));
+                                }
+
+                                // Determine the AND array size
+                                numBytesPerLine = checked((((bih.biWidth) + 31) & ~31) >> 3);
+                                int andSize = checked(numBytesPerLine * iconHeight);
+                                iidata.iconAND = new byte[andSize];
+                                nread = bihReader.Read(iidata.iconAND, 0, andSize);
+                                if (nread != andSize)
+                                {
+                                    string msg = string.Format("{0} data length expected {1}, read {2}", "AND", andSize, nread);
+                                    throw new ArgumentException(msg, nameof(stream));
+                                }
+                            }
+                            catch (OverflowException)
+                            {
+                                throw new ArgumentException(nameof(stream));
+                            }
 
-                //is this is the best fit??
-                if (!sizeObtained)
-                {
-                    if (((ide.height == height) || (ide.width == width)) && !ide.ignore)
-                    {
-                        this.id = (ushort)i;
-                        sizeObtained = true;
-                        this.iconSize.Height = ide.height;
-                        this.iconSize.Width = ide.width;
+                            imageData[i] = iidata;
+
+                            // is this is the best fit?
+                            if (!sizeObtained)
+                            {
+                                if (((ide.height == height) || (ide.width == width)))
+                                {
+                                    this.id = (ushort)i;
+                                    this.iconSize.Height = ide.height;
+                                    this.iconSize.Width = ide.width;
+                                    sizeObtained = true;
+                                }
+                            }
+                        }
                     }
-                }
-            }
 
-            // throw error if no valid entries found
-            int valid = 0;
-            for (int i = 0; i < dirEntryCount; i++)
-            {
-                if (!(iconDir.idEntries[i].ignore))
-                    valid++;
+                    stream.Seek(streamPosition, SeekOrigin.Begin);
+                }
             }
 
-            if (valid == 0)
+            if (!valid)
                 throw new Win32Exception(0, "No valid icon entry were found.");
```

Comment:
This should be in the resources

Reply:
Or it could do what Icon.Windows.cs does
```c#
                if (_bestBytesInRes < 0)
                {
                    throw new Win32Exception(SafeNativeMethods.ERROR_INVALID_PARAMETER);
                }
```